thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/detection/components/WoWStdRuleDetector.java

```diff
@@ -0,0 +1,275 @@
+    for (int k = 0; k < size; k++) {
+      DataFrame trainingDF;
+      trainingDF = getLookbackDF(inputDF, forecastDF.getLong(COL_TIME, k));
+      //the get historical WoW mean and std.
+      std[k]= trainingDF.getDoubles(COL_CHANGE).std().value();
+      mean[k] = trainingDF.getDoubles(COL_CHANGE).mean().value();
+
+      // We need at least 4 weeks of data
+      if (trainingDF.size() < 4) {
+        continue;
+      }
+
+      //calculate baseline, error , upper and lower bound for prediction window.
+      resultTimeArray[k] = forecastDF.getLong(COL_TIME, k);
+      baselineArray[k] = trainingDF.getDouble(COL_VALUE,trainingDF.size()-1) * (1 + mean[k]);
+      errorArray[k] = trainingDF.getDouble(COL_VALUE,trainingDF.size()-1) * sensitivity * std[k];
+      upperBoundArray[k] = baselineArray[k] + errorArray[k];
+      lowerBoundArray[k] = baselineArray[k] - errorArray[k];
+    }
+    //Construct the dataframe.
+    resultDF.addSeries(COL_TIME, LongSeries.buildFrom(resultTimeArray)).setIndex(COL_TIME);
+    resultDF.addSeries(COL_VALUE, DoubleSeries.buildFrom(baselineArray));
+    resultDF.addSeries(COL_UPPER_BOUND, DoubleSeries.buildFrom(upperBoundArray));
+    resultDF.addSeries(COL_LOWER_BOUND, DoubleSeries.buildFrom(lowerBoundArray));
+    resultDF.addSeries(COL_ERROR, DoubleSeries.buildFrom(errorArray));
+    return resultDF;
+  }
+
+  /**
+   * Fetch data from metric
+   *
+   * @param metricEntity metric entity
+   * @param start start timestamp
+   * @param end end timestamp
+   * @return Data Frame that has data from start to end
+   */
+  private DataFrame fetchData(MetricEntity metricEntity, long start, long end) {
+    List<MetricSlice> slices = new ArrayList<>();
+    MetricSlice sliceData = MetricSlice.from(metricEntity.getId(), start, end,
+        metricEntity.getFilters(), timeGranularity);
+    slices.add(sliceData);
+    LOG.info("Getting data for" + sliceData.toString());
+    InputData data = this.dataFetcher.fetchData(new InputDataSpec().withTimeseriesSlices(slices)
+        .withMetricIdsForDataset(Collections.singletonList(metricEntity.getId())));
+    return data.getTimeseries().get(sliceData);
+  }
+
+  /**
+   * Returns a data frame containing lookback number of data before prediction time
+   * @param originalDF the original dataframe
+   * @param time the prediction time, in unix timestamp
+   * @return DataFrame containing lookback number of data
+   */
+
+  private DataFrame getLookbackDF(DataFrame originalDF, Long time) {
+    LongSeries longSeries = (LongSeries) originalDF.get(COL_TIME);
+    int indexFinish = longSeries.find(time);
+    DataFrame df = DataFrame.builder(COL_TIME, COL_VALUE).build();
+
+    if (indexFinish != -1) {
+      int indexStart = Math.max(0, indexFinish - lookback);
+      df = df.append(originalDF.slice(indexStart, indexFinish));
+    }
+    // calculate percentage change
+    df.addSeries(COL_CURR,df.getDoubles(COL_VALUE).shift(-1));
```

Comment:
The `originalDF` might not be weekly granularity. Looks like it's set by the `monitoringGranularity`. Is this calculating WoW percentage change? Shifting by 1 might not get the correct result.

If we are always expecting the `originalDF` to be `7_DAYS` granularity, maybe we shouldn't let it be configurable in the specs.

Reply:
The `originalDF` could be weekly, monthly, quarterly(given enough data) or any period set by the user. I think it does not make sense for this detection component to look at DoD change if there is seasonality. However, if the metric is super stable, it can also use DoD change to predict too.